lib/Dialect/OM/OMOps.cpp

```diff
@@ -415,25 +424,15 @@ void circt::om::ClassOp::addNewFieldsOp(mlir::OpBuilder &builder,
   }
   // Also store the locations incase there's some other analysis that might
   // be able to use the default FusedLoc representation.
-  builder.create<ClassFieldsOp>(
-      builder.getFusedLoc(locs, builder.getArrayAttr(locAttrs)), values);
+  builder.create<ClassFieldsOp>(builder.getFusedLoc(locs), values,
+                                builder.getArrayAttr(locAttrs));
 }
 
 mlir::Location circt::om::ClassOp::getFieldLocByIndex(size_t i) {
-  Location loc = this->getFieldsOp()->getLoc();
-  if (auto locs = dyn_cast<FusedLoc>(loc)) {
-    // Because it's possible for a user to construct a fields op directly and
-    // place a FusedLoc that doersn't follow the storage format of
-    // addNewFieldsOp, we assert the information has been stored appropriately
-    ArrayAttr metadataArr = dyn_cast<ArrayAttr>(locs.getMetadata());
-    assert(metadataArr && "Expected fused loc to store metadata array");
-    assert(i < metadataArr.size() &&
-           "expected index to be less than array size");
-    LocationAttr locAttr = dyn_cast<LocationAttr>(metadataArr[i]);
-    assert(locAttr && "expected metadataArr entry to be location attribute");
-    loc = Location(locAttr);
-  }
-  return loc;
+  auto fieldLocs = this->getFieldsOp().getFieldsLocs();
+  if (!fieldLocs)
+    return this->getFieldsOp().getLoc();
+  return cast<LocationAttr>(fieldLocs.value()[i]);
```

Comment:
This needs to check the array is the right size before doing this indexing. Should probably put that in a verifier on the operation, which checks the locations size matches the fields size, if the locations array is specified. Also see my other comment--I'm wondering if this should just be an inherent, non-optional attribute... but either way, we should probably verify the size in a verifier at least.

Reply:
Makes sense, yea I think since we likely want to omit the location info unless the flag is set, implementing a custom parse/print will allow us to enforce the invariant that the array exists and is the right size.  We would want to verify that no one changes that in a pass.

Reply:
Fair enough, I'm mainly thinking we should have a check before we index into the array.